tools/defrag_mem_test.py

```diff
@@ -0,0 +1,151 @@
+    number of keys: 4,023,467
+    value size: 64 bytes
+    key name pattern: key-for-testing
+    host: localhost
+    port: default redis port
+    Please note that this would create 4 * number of keys entries
+    You can see the memory usage/defrag state with the monitoring task that
+    prints the current state
+
+NOTE:
+    If this seems to get stuck please kill it with ctrl+c
+    This can happen in case we don't have "defrag_realloc_total > 0"
+'''
+
+class TaskCancel:
+    def __init__(self):
+        self.run = True
+
+    def dont_stop(self):
+        return self.run
+
+    def stop(self):
+        self.run = False
+
+async def run_cmd(connection, cmd, sub_val):
+    val = await connection.execute_command(cmd, sub_val)
+    return val
+
+async def stats_check(connection, condition):
+    try:
+        defrag_task_invocation_total = 0;
+        runs=0
+        while condition.dont_stop():
+            await asyncio.sleep(5)
+            runs += 1
+            info = await run_cmd(connection, "info", "stats")
+            if info is not None:
+                if info['defrag_task_invocation_total'] != defrag_task_invocation_total:
+                    print("--------------------------------------------------------------")
+                    print(f"defrag_task_invocation_total: {info['defrag_task_invocation_total']:,}")
+                    print(f"defrag_realloc_total: {info['defrag_realloc_total']:,}")
+                    print(f"defrag_attempt_total: {info['defrag_attempt_total']:,}")
+                    print("--------------------------------------------------------------")
+                    if info["defrag_realloc_total"] > 0:
+                        print("defrag task successfully found memory locations to reallocate")
+                        condition.stop()
+                        return True
+                    defrag_task_invocation_total = info['defrag_task_invocation_total']
+            if runs % 12 == 0:
+                print("--------------------------------------------------------------")
+                info = await run_cmd(connection, "info", "memory")
+                print(f"memory commited: {info['comitted_memory']:,}")
+                print(f"memory used: {info['used_memory']:,}")
+                print(f"memory usage ratio: {info['comitted_memory']/info['used_memory']:.2f}")
+                print("--------------------------------------------------------------")
+    except Exception as e:
+        print(f"failed to run monitor task: {e}")
+    return False
+
+
+async def delete_keys(connection, keys):
+    #pipe = connection.pipeline()
+    #for key in keys:
+    #    pipe.delete(key)
+    #print(f"deleting {len(keys)} keys")
+    results = await connection.delete(*keys)
+    #pipe.execute()
+    return results
+
+async def mem_cleanup(connection, pattern, num, cond):
+    cursor=0
+    while cond.dont_stop():
+        cur, keys = await  connection.scan(cursor=cursor, match=pattern, count=100)
```

Comment:
why do you need to scan them? you already know which keys are there.
so if you generate keys `foo:0` - `foo:100000`, you can relatively quickly generate inside python 
keys `foo:1*` that fit this range (just by going over the range and filtering out those that do not fit).
I think it will be faster than scanning.

Reply:
I was thinking about it, Adi told me it tied this operation with the way populate works. I think I will go generating keys the way I did in the unit test

Reply:
that's true but you said it's slow due to erases.

Reply:
I think convenience here is more important because I do not see how populate will change drastically in the near term